src/blocking/spi.rs

```diff
@@ -102,3 +102,52 @@ pub mod write_iter {
         }
     }
 }
+
+/// Operation for transactional SPI trait
+///
+/// This allows composition of SPI operations into a single bus transaction
+#[derive(Debug, PartialEq)]
+pub enum Operation<'a, W: 'static> {
+    /// Write data from the provided buffer, discarding read data
+    Write(&'a [W]),
+    /// Write data out while reading data into the provided buffer
+    Transfer(&'a mut [W]),
+}
+
+/// Transactional trait allows multiple actions to be executed
+/// as part of a single SPI transaction
+pub trait Transactional<W: 'static> {
+    /// Associated error type
+    type Error;
+
+    /// Execute the provided transactions
+    fn try_exec<'a>(&mut self, operations: &mut [Operation<'a, W>]) -> Result<(), Self::Error>;
+}
+
+/// Blocking transactional impl over spi::Write and spi::Transfer
+pub mod transactional {
+    use super::{Operation, Transfer, Write};
+
+    /// Default implementation of `blocking::spi::Transactional<W>` for implementers of
+    /// `spi::Write<W>` and `spi::Transfer<W>`
+    pub trait Default<W>: Write<W> + Transfer<W> {}
+
+    impl<W: 'static, E, S> super::Transactional<W> for S
+    where
+        S: self::Default<W> + Write<W, Error = E> + Transfer<W, Error = E>,
+        W: Copy + Clone,
+    {
+        type Error = E;
+
+        fn try_exec<'a>(&mut self, operations: &mut [super::Operation<'a, W>]) -> Result<(), E> {
+            for op in operations {
+                match op {
+                    Operation::Write(w) => self.try_write(w)?,
+                    Operation::Transfer(t) => self.try_transfer(t).map(|_| ())?,
+                }
+            }
```

Comment:
I guess this works fine because SPI does not have an end-of-transmission marker like I2C's STOP.
However, would it be a problem if the `try_write` takes too long to end or `try_transfer` takes too long to start? (for whatever definition of "too long").
Do you think it is correct to think that this _in-between_ time would just count as holding the bus like with I2C's clock-stretching (since the clock signal is just stopped); or do you know of any other timeouts in the MCU HAL implementation or so that may "break up the transaction"?

Reply:
yeah this is fine because while you're not writing things you're not outputting the clock signal and a transaction is grouped by the chip select, so it's just like the gap between any two bytes in normal operation*. it's also exactly how most SPI drivers work _anyway_ (assert CS, transfer to set reg address / read state, write/read the data from this address, de-assert CS).

\* it might be a slightly different length but for all intents and purposes it's fine.

Reply:
Sounds good. I wanted to highlight it just in case. Thanks!